nntrainer/include/bn_layer.h

```diff
@@ -0,0 +1,151 @@
+#include <iostream>
+#include <optimizer.h>
+#include <tensor.h>
+#include <vector>
+#include <layer.h>
+
+namespace nntrainer {
+
+/**
+ * @class   BatchNormalizationLayer
+ * @brief   Batch Noramlization Layer
+ */
+class BatchNormalizationLayer : public Layer {
+public:
+  /**
+   * @brief     Constructor of Batch Noramlization Layer
+   */
+  BatchNormalizationLayer() : epsilon(0.0){};
+
+  /**
+   * @brief     Destructor of BatchNormalizationLayer
+   */
+  ~BatchNormalizationLayer(){};
+
+  /**
+   * @brief     Read Weight & Bias Data from file
+   * @param[in] file input stream file
+   */
+  void read(std::ifstream &file);
+
+  /**
+   * @brief     Save Weight & Bias Data to file
+   * @param[in] file output stream file
+   */
+  void save(std::ofstream &file);
+
+  /**
+   * @brief     forward propagation with input
+   * @param[in] in Input Tensor from upper layer
+   * @retval    Activation(W x input + B)
+   */
+  Tensor forwarding(Tensor in, int &status);
+
+  /**
+   * @brief     foward propagation : return Input Tensor
+   *            It return Input as it is.
+   * @param[in] input input Tensor from lower layer.
+   * @param[in] output label Tensor.
+   * @retval    Activation(W x input + B)
+   */
+  Tensor forwarding(Tensor in, Tensor output, int &status) {
+    return forwarding(in, status);
+  };
+
+  /**
+   * @brief     back propagation
+   *            Calculate dJdB & dJdW & Update W & B
+   * @param[in] in Input Tensor from lower layer
+   * @param[in] iteration Number of Epoch for ADAM
+   * @retval    dJdB x W Tensor
+   */
+  Tensor backwarding(Tensor in, int iteration);
+
+  /**
+   * @brief     set optimizer
+   * @param[in] opt Optimizer
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  int setOptimizer(Optimizer &opt);
+
+  /**
+   * @brief     copy layer
+   * @param[in] l layer to copy
+   */
+  void copy(std::shared_ptr<Layer> l);
+
+  /**
+   * @brief     initialize layer
+   * @param[in] b batch size
+   * @param[in] h height
+   * @param[in] w width
+   * @param[in] last last layer
+   * @param[in] init_zero boolean to set Bias zero
+   * @param[in] wini Weight Initialization Scheme
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  int initialize(int b, int h, int w, bool last, bool init_zero,
+                 WeightIniType wini);
+
+  /**
+   * @brief     set Property of layer
+   * @param[in] values values of property
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  int setProperty(std::vector<std::string> values);
+
+  /**
+   * @brief     Property Enumeration
+   *            0. input shape : string
+   *            1. bias zero : bool
+   *            5. epsilon : float
+   */
+  enum class PropertyType {
+    input_shape = 0,
+    bias_zero = 1,
+    epsilon = 5,
+  };
+
+private:
+  Tensor weight;
```

Comment:
I think bn layer dont need weight or bias.

Reply:
yes. bn layer needs verification deeply with test cases and application. But not this PR I think.

Reply:
ok.